frame/election-provider-multi-phase/src/weights.rs

```diff
@@ -0,0 +1,150 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Autogenerated weights for pallet_election_provider_multi_phase
+//!
+//! THIS FILE WAS AUTO-GENERATED USING THE SUBSTRATE BENCHMARK CLI VERSION 3.0.0
+//! DATE: 2021-02-12, STEPS: [50, ], REPEAT: 20, LOW RANGE: [], HIGH RANGE: []
+//! EXECUTION: Some(Wasm), WASM-EXECUTION: Compiled, CHAIN: Some("dev"), DB CACHE: 128
+
+// Executed Command:
+// target/release/substrate
+// benchmark
+// --chain=dev
+// --steps=50
+// --repeat=20
+// --pallet=pallet_election_provider_multi_phase
+// --extrinsic=*
+// --execution=wasm
+// --wasm-execution=compiled
+// --heap-pages=4096
+// --output=./frame/election-provider-multi-phase/src/weights.rs
+// --template=./.maintain/frame-weight-template.hbs
+
+
+#![allow(unused_parens)]
+#![allow(unused_imports)]
+
+use frame_support::{traits::Get, weights::{Weight, constants::RocksDbWeight}};
+use sp_std::marker::PhantomData;
+
+/// Weight functions needed for pallet_election_provider_multi_phase.
+pub trait WeightInfo {
+	fn on_initialize_nothing() -> Weight;
+	fn on_initialize_open_signed() -> Weight;
+	fn on_initialize_open_unsigned_with_snapshot() -> Weight;
+	fn on_initialize_open_unsigned_without_snapshot() -> Weight;
+	fn submit_unsigned(v: u32, t: u32, a: u32, d: u32, ) -> Weight;
+	fn feasibility_check(v: u32, t: u32, a: u32, d: u32, ) -> Weight;
+}
+
+/// Weights for pallet_election_provider_multi_phase using the Substrate node and recommended hardware.
+pub struct SubstrateWeight<T>(PhantomData<T>);
+impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
+	fn on_initialize_nothing() -> Weight {
+		(23_272_000 as Weight)
+			.saturating_add(T::DbWeight::get().reads(7 as Weight))
```

Comment:
FYI most of these reads are in the `predict_next_election` from staking's side... This is one of the aspects of this PR that I am still not super happy with and gladly accept calls to rethink it: how we predict when next election happens.

Reply:
Also, we do this every block, maybe we can do it better?

Reply:
yes, but also, at some point, if they are read everyblock, we can consider them always in cache no ?

Reply:
They are read only by me afaik, so sadly I think they are not cached in any way. 

Noting that we don't have any cache that caches stuff from this block to the next one right? it is only intra-block caching.

Reply:
not sure, but there can be a cache inter block, for reads too, in `client/db/src/storage_cache.rs` maybe @cheme knows about it.

Reply:
yes, 'client/db/src/storage_cache.rs' is caching between consecutive blocks, got reset if there is a reorg though, and it's lru with client specific size so can't really have expectation.

Reply:
I remember thinking on how to avoid redundant calculation between blocks, and thought of something like: `calculated_cached_value(cache_key, source: [storage_keys | storage_prefix], fn calculate(source)-> result) -> result` that would either:
- take a cached result from a previous block if no source value did change and a cache is implemented (this requires to invalidate cache on any key access from state-machine (with a radix trie filter for instance))
- calculate the result from the function 'calculate', then return it and cache it (if such cache is implemented).
But this would require new host function to access cache and invalidate on source storage access.

But tbh it is not very easy to implement and is only useful for costly computation.
But in case of costly computation we can also simply write the result in state and have the invalidate logic coded in the runtime. It is possibly less efficient but since it will behave the same between client, calculated weight can benefit from it.